File: packages/server/src/engine/reranker.ts
/**
 * Second-pass cross-encoder reranker — Mem0 alignment, Phase 5.
 *
 * EXPERIMENT, behind a flag (config `rerank.enabled` + per-request
 * `rerank: true`). Adopted only if it beats the Phase 4 configuration on
 * answer accuracy at n≥50; otherwise this file and its wiring are
 * deleted, not left dormant (plan §4, process rule 7 — our only prior
 * measurement of a cross-encoder was harmful, and Mem0's evidence is not
 * our workload).
 *
 * Shape: the engine overfetches a candidate pool (poolMultiplier × k),
 * this class scores every (query, document) pair with a cross-encoder
 * service, and the engine re-orders the pool by that score before the
 * existing diversity + token-budget selection runs. Wire format is the
 * Jina/Cohere-compatible `/rerank` contract that vLLM and TEI both
 * serve for models like BAAI/bge-reranker-v2-m3:
 *
 *   POST <endpoint>  {model, query, documents: string[]}
 *   → {results: [{index, relevance_score}, …]}
 */

export interface SearchRerankerConfig {
  /** Full URL of the rerank endpoint (no path is appended). */
  endpoint: string;
  model: string;
  apiKey?: string;
  timeoutMs: number;
}

/** Documents are clipped before shipping: a cross-encoder scores the
 *  pair on its first window anyway, and an unclipped conversation chunk
 *  can be 24KB — pool × that is real bandwidth on every flagged search. */
const MAX_DOC_CHARS = 2000;

interface RerankResponse {
  results?: Array<{ index?: number; relevance_score?: number }>;
}

export class SearchReranker {
  constructor(private readonly cfg: SearchRerankerConfig) {}

  /** Score `documents` against `query`. Returns a score per input index
   *  (missing entries stay undefined so the caller can keep the fused
   *  order for anything the service did not score). Throws on transport
   *  or contract errors — the caller decides that a failed rerank means
   *  "fall back to fused order", not "fail the search". */
  async rerank(query: string, documents: string[]): Promise<Array<number | undefined>> {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), this.cfg.timeoutMs);
    try {
      const res = await fetch(this.cfg.endpoint, {
        method: "POST",
        headers: {
          "content-type": "application/json",
          ...(this.cfg.apiKey ? { authorization: `Bearer ${this.cfg.apiKey}` } : {}),
        },
        body: JSON.stringify({
          model: this.cfg.model,
          query,
          documents: documents.map((d) => d.slice(0, MAX_DOC_CHARS)),
        }),
        signal: ctrl.signal,
      });
      if (!res.ok) {
        throw new Error(`rerank endpoint ${res.status}: ${(await res.text()).slice(0, 200)}`);
      }
      const body = (await res.json()) as RerankResponse;
      if (!Array.isArray(body.results)) {
        throw new Error("rerank response missing results[]");
      }
      const scores: Array<number | undefined> = new Array(documents.length);
      for (const r of body.results) {
        if (
          typeof r.index === "number" &&
          r.index >= 0 &&
          r.index < documents.length &&
          typeof r.relevance_score === "number"
        ) {
          scores[r.index] = r.relevance_score;
        }
      }
      return scores;
    } finally {
      clearTimeout(timer);
    }
  }
}

File: go/internal/embeddings/reranker.go
// Second-pass cross-encoder reranker client. Transcribed from
// packages/server/src/engine/reranker.ts (read-only reference). Wire
// format is the Jina/Cohere-compatible /rerank contract vLLM and TEI
// serve for models like BAAI/bge-reranker-v2-m3:
//
//	POST <endpoint>  {model, query, documents: string[]}
//	→ {results: [{index, relevance_score}, …]}
package embeddings

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"time"
)

// maxDocChars — reranker.ts MAX_DOC_CHARS: documents are clipped before
// shipping; a cross-encoder scores the first window anyway.
const maxDocChars = 2000

type RerankerConfig struct {
	Endpoint  string // full URL — no path is appended
	Model     string
	APIKey    string
	TimeoutMs int
}

type Reranker struct {
	cfg  RerankerConfig
	http *http.Client
}

func NewReranker(cfg RerankerConfig) *Reranker {
	return &Reranker{cfg: cfg, http: &http.Client{}}
}

// Rerank scores documents against query. The result has one slot per
// input index; nil slots were not scored by the service (the caller
// keeps the fused order for those). Errors are transport/contract
// failures — the caller falls back to fused order, never fails the
// search (reranker.ts contract).
func (r *Reranker) Rerank(ctx context.Context, query string, documents []string) ([]*float64, error) {
	clipped := make([]string, len(documents))
	for i, d := range documents {
		clipped[i] = utf16Slice(d, maxDocChars)
	}
	body, err := json.Marshal(map[string]any{
		"model":     r.cfg.Model,
		"query":     query,
		"documents": clipped,
	})
	if err != nil {
		return nil, err
	}
	reqCtx, cancel := context.WithTimeout(ctx, time.Duration(r.cfg.TimeoutMs)*time.Millisecond)
	defer cancel()
	req, err := http.NewRequestWithContext(reqCtx, http.MethodPost, r.cfg.Endpoint, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("content-type", "application/json")
	if r.cfg.APIKey != "" {
		req.Header.Set("authorization", "Bearer "+r.cfg.APIKey)
	}
	res, err := r.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close() //nolint:errcheck
	if res.StatusCode != http.StatusOK {
		detail, _ := io.ReadAll(io.LimitReader(res.Body, 200))
		return nil, fmt.Errorf("rerank endpoint %d: %s", res.StatusCode, string(detail))
	}
	var parsed struct {
		Results []struct {
			Index          *float64 `json:"index"`
			RelevanceScore *float64 `json:"relevance_score"`
		} `json:"results"`
	}
	if err := json.NewDecoder(res.Body).Decode(&parsed); err != nil {
		return nil, err
	}
	if parsed.Results == nil {
		return nil, fmt.Errorf("rerank response missing results[]")
	}
	scores := make([]*float64, len(documents))
	for _, item := range parsed.Results {
		if item.Index == nil || item.RelevanceScore == nil {
			continue
		}
		idx := *item.Index
		// Integer, in range, finite score — reranker.ts validation.
		if idx != math.Trunc(idx) || idx < 0 || int(idx) >= len(documents) {
			continue
		}
		if math.IsNaN(*item.RelevanceScore) || math.IsInf(*item.RelevanceScore, 0) {
			continue
		}
		s := *item.RelevanceScore
		scores[int(idx)] = &s
	}
	return scores, nil
}
